pkg/transport/custom_transform_collection.go

```diff
@@ -0,0 +1,244 @@
+	// It is used to get the CustomTransform objects relevant to a given GroupResource.
+	getTransformObjects func(indexName, indexedValue string) ([]any, error)
+
+	// enqueue is used to add a reference to a Binding that needs to be re-processed because
+	// of a change to a CustomTransform that the Binding is sensitive to.
+	enqueue func(any)
+
+	// mutex must be locked while accessing the following fields or their contents.
+	// The comments on the following fields and on groupResourceTransformData
+	// are things that hold true while the mutex is not locked.
+	mutex sync.Mutex
+
+	// grToTransformData has an entry for every GroupResource that some Binding cares about
+	// (i.e., lists an object of that GroupResource), and no more entries.
+	grToTransformData map[metav1.GroupResource]*groupResourceTransformData
+
+	// ctNameToSpec holds, for each CustomTranform whose spec contributed to an
+	// entry in grToTransformData, that CustomTransformSpec.
+	ctNameToSpec map[string]v1alpha1.CustomTransformSpec
+
+	// bindingNameToGroupResources tracks the set of GroupResource that each Binding
+	// references. This is so that when the set for a given Binding changes,
+	// for the GroupResources that are no longer in the set, the Binding's Name can
+	// be removed from groupResourceTransformData.bindingsThatCare.
+	// No Set[GroupResource] here is empty.
+	bindingNameToGroupResources map[string]sets.Set[metav1.GroupResource]
+}
+
+// groupResourceTransformData is the ingested custom transforms for a given GroupResource
+type groupResourceTransformData struct {
+	bindingsThatCare sets.Set[string /*Binding name*/] // not empty
+	removes          []jsonpath.Query
+}
+
+func newCustomTransformCollection(client controlclient.CustomTransformInterface, getTransformObjects func(indexName, indexedValue string) ([]any, error), enqueue func(any)) *customTransformCollection {
+	return &customTransformCollection{
+		client:                      client,
+		getTransformObjects:         getTransformObjects,
+		enqueue:                     enqueue,
+		grToTransformData:           make(map[metav1.GroupResource]*groupResourceTransformData),
+		ctNameToSpec:                make(map[string]v1alpha1.CustomTransformSpec),
+		bindingNameToGroupResources: make(map[string]sets.Set[metav1.GroupResource]),
+	}
+}
+
+// GetCustomTransformData returns the groupResourceTransformData to use
+// for the given GroupResource on behalf of the named Binding.
+// This method returns a cached answer if one is available, otherwise
+// digests the relevant CustomTransform object(s) and caches the result.
+// Always records the fact that the given binding depends on the answer.
+func (ctc *customTransformCollection) GetCustomTransformData(ctx context.Context, groupResource metav1.GroupResource, bindingName string) *groupResourceTransformData {
+	logger := klog.FromContext(ctx)
+	ctc.mutex.Lock()
+	defer ctc.mutex.Unlock()
+	grtd, ok := ctc.grToTransformData[groupResource]
```

Comment:
`grtd` constantly requires translation in head while reading, which makes building understanding harder - what do you think about `grTransformData`? Other variable names seem to follow similar structures.

Reply:
OK, renamed.